server/repository/src/db_diesel/temperature_chart/temperature_chart.rs

```diff
@@ -0,0 +1,355 @@
+use crate::{
+    db_diesel::temperature_log_row::temperature_log::dsl as temperature_log_dsl, DBType,
+    RepositoryError, StorageConnection, TemperatureChartRow, TemperatureLogFilter,
+    TemperatureLogRepository,
+};
+use chrono::NaiveDateTime;
+use diesel::prelude::*;
+use util::datetime_with_offset;
+
+use super::temperature_chart_row::*;
+
+pub struct TemperatureChartRepository<'a> {
+    connection: &'a StorageConnection,
+}
+
+type QueryResult = (NaiveDateTime, NaiveDateTime, f64, String);
+
+impl<'a> TemperatureChartRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        TemperatureChartRepository { connection }
+    }
+
+    pub fn query(
+        &self,
+        from_datetime: NaiveDateTime,
+        to_datetime: NaiveDateTime,
+        number_of_data_points: i32,
+        temperature_log_filter: Option<TemperatureLogFilter>,
+    ) -> Result<(Vec<TemperatureChartRow>, Vec<Interval>), RepositoryError> {
+        let intervals = calculate_intervals(from_datetime, to_datetime, number_of_data_points);
+        let mut query = TemperatureChart {
+            intervals: intervals.clone(),
+        }
+        .into_boxed::<DBType>();
+
+        if temperature_log_filter.is_some() {
+            let temperature_log_ids =
+                TemperatureLogRepository::create_filtered_query(temperature_log_filter)
+                    .select(temperature_log_dsl::id);
+            query = query.filter(TemperatureLogId.eq_any(temperature_log_ids));
+        };
+
+        let query = query
+            .select((FromDatetime, ToDatetime, AverageTemperature, SensorId))
+            .group_by((FromDatetime, ToDatetime, SensorId));
+
+        // First by sensor then by datetime (so should be sorted by sensor and then by datetime)
+        let query = query.order_by((SensorId.asc(), FromDatetime.asc()));
+
+        // Debug diesel query
+        // println!("{}", diesel::debug_query::<DBType, _>(&query).to_string());
+
+        let chart_data = query
+            .load::<QueryResult>(&self.connection.connection)?
+            .into_iter()
+            .map(TemperatureChartRow::from)
+            .collect();
+
+        Ok((chart_data, intervals))
+    }
+}
+
+impl From<QueryResult> for TemperatureChartRow {
+    fn from((from_datetime, to_datetime, average_temperature, sensor_id): QueryResult) -> Self {
+        Self {
+            from_datetime,
+            to_datetime,
+            average_temperature,
+            sensor_id,
+        }
+    }
+}
+
+fn calculate_intervals(
+    from_datetime: NaiveDateTime,
+    to_datetime: NaiveDateTime,
+    number_of_data_points: i32,
+) -> Vec<Interval> {
+    let interval = (to_datetime - from_datetime) / number_of_data_points;
```

Comment:
Technically if there is a temperature log on to_datetime, it won't be included in the chart aggregation, should it ?

Reply:
> not in that interval no;
> should be temperature_log.datetime >= from_datetime && temperature_log.datetime < to_datetime

Great, I think that's the current logic